### PR TITLE
tablets: add warning on CREATE KEYSPACE

### PR DESCRIPTION
**Pull request by nyh**
The CDC feature is not supported on a table that uses tablets
(Refs https://github.com/scylladb/scylladb/issues/16317), so if a user creates a keyspace with tablets enabled
they may be surprised later (perhaps much later) when they try to enable
CDC on the table and can't.

The LWT feature always had issue Refs https://github.com/scylladb/scylladb/issues/5251, but it has become potentially
more common with tablets.

So it was proposed that as long as we have missing features (like CDC or
LWT), every time a keyspace is created with tablets it should output a
warning (a bona-fide CQL warning, not a log message) that some features
are missing, and if you need them you should consider re-creating the
keyspace without tablets.

This PR does this.

The warning text which will be produced is the following (obviously, it can
be improved later, as we perhaps find more missing features):

>   "Tables in this keyspace will be replicated using tablets, and will
>    not support the CDC feature (issue https://github.com/scylladb/scylladb/issues/16317) and LWT may suffer from
>    issue https://github.com/scylladb/scylladb/issues/5251 more often. If you want to use CDC or LWT, please drop
>    this keyspace and re-create it without tablets, by adding AND TABLETS
>    = {'enabled': false} to the CREATE KEYSPACE statement."

This PR also includes a test - that checks that this warning is is
indeed generated when a keyspace is created with tablets (either by default
or explicitly), and not generated if the keyspace is created without
tablets. It also fixes existing tests which didn't like the new warning.

Fixes https://github.com/scylladb/scylladb/issues/16807

### CODE REVIEW

**Review by denesb — APPROVED** [review]

**Comment by nyh** [comment]
Hmm, this is an actual test failure caused by my change.
The cql-pytest `test_guardrail_replication_strategy.py`::`test_given_default_config_when_creating_ks_should_only_produce_warning_for_simple_strategy` wants to check if guardrails add warnings, but assumes there is no other warnings - so it wants to see 0 or 1 warnings. But in this PR, we add another warning to every keyspace creation, so this test breaks.

I'll need to modify this test to look for a specific guardrails message in the warnings, and not assume that there can't be any other warnings.

**Comment by nyh** [comment]
Pushed a new version that:
1. Fixes failing guardrails test which didn't like the new warning
2. Add LWT to the warning text (as requested by @avikivity. CC @kostja @bhalevy).

File: cql3/statements/create_keyspace_statement.cc [hunk 1]
```diff
@@ -94,21 +94,42 @@ future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector
...
         ret = ::make_shared<event::schema_change>(
                 event::schema_change::change_type::CREATED,
                 event::schema_change::target_type::KEYSPACE,
                 keyspace());
+
+        // If the new keyspace uses tablets, as long as there are features
+        // which aren't supported by tablets we want to warn the user that
+        // they will not be usable on the new keyspace - and suggest how a
+        // keyspace can be created without tablets (see rationale in #16807).
+        // Once all feature will become supported with tablets, we should
```
Comment (ptrsmrn) [inline cql3/statements/create_keyspace_statement.cc:112]:
```suggestion
        // Once all features will become supported with tablets, we should
```

File: cql3/statements/create_keyspace_statement.cc [hunk 2]
```diff
@@ -94,21 +94,42 @@ future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector
...
+        // remove this check.
+        auto rs = locator::abstract_replication_strategy::create_replication_strategy(
+            ksm->strategy_name(),
+            locator::replication_strategy_params(ksm->strategy_options(), ksm->initial_tablets()));
+        if (rs->uses_tablets()) {
+            warnings.push_back(
+                "Tables in this keyspace will be replicated using tablets, "
```
Comment (ptrsmrn) [inline cql3/statements/create_keyspace_statement.cc:119]: "Tables ... tablets..." - these 2 words are so similar, it's confusing. Since it's a message shown to the client, maybe we should emphasize what we really mean. But that's just a food for thought.

File: cql3/statements/create_keyspace_statement.cc [hunk 3]
```diff
@@ -94,21 +94,42 @@ future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector
...
+                "and will not support the CDC feature (issue #16317) and LWT "
+                "may suffer from issue #5251 more often. If you want to use "
+                "CDC or LWT, please drop this keyspace and re-create it "
+                "without tablets, by adding AND TABLETS = {'enabled': false} "
```
Comment (ptrsmrn) [inline cql3/statements/create_keyspace_statement.cc:123]: `verbatim` on CQL could prove more readable.
```suggestion
                "without tablets, by adding `AND TABLETS = {'enabled': false}` "
```

File: test/cql-pytest/test_guardrail_replication_strategy.py [hunk 4]
```diff
@@ -20,20 +20,26 @@ def create_ks_and_assert_warnings_and_errors(cql, ks_opts, warnings_count=0, fai
         response_future = cql.execute_async("CREATE KEYSPACE " + keyspace + " " + ks_opts)
         response_future.result()
         cql.execute("DROP KEYSPACE " + keyspace)
-        if warnings_count:
-            assert response_future.warnings is not None and len(response_future.warnings) == warnings_count
-        else:
-            assert response_future.warnings is None
+
+        if warnings_count is not None:
```
Comment (ptrsmrn) [inline test/cql-pytest/test_guardrail_replication_strategy.py:24]: That if-ology, although correct, may prove a bit hard to follow for our future readers, maybe some extra explanatory comment?

File: cql3/statements/create_keyspace_statement.cc [hunk 5]
```diff
@@ -94,21 +94,42 @@ future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector
...
+        // keyspace can be created without tablets (see rationale in #16807).
+        // Once all feature will become supported with tablets, we should
+        // remove this check.
+        auto rs = locator::abstract_replication_strategy::create_replication_strategy(
+            ksm->strategy_name(),
+            locator::replication_strategy_params(ksm->strategy_options(), ksm->initial_tablets()));
+        if (rs->uses_tablets()) {
+            warnings.push_back(
+                "Tables in this keyspace will be replicated using tablets, "
+                "and will not support the CDC feature (issue #16317) and LWT "
```
Comment (ptrsmrn) [inline cql3/statements/create_keyspace_statement.cc:120]: CDC & LWT are mentioned explicitly, but aren't/won't there be any other such features? Maybe we can generalize this statement.

File: cql3/statements/create_keyspace_statement.cc (hunk 2 above)
Comment (nyh) [reply cql3/statements/create_keyspace_statement.cc:119]: I know, and I often fall on this confusion myself, but this is the word that Avi decided to use.... "vnodes" was a better word in the sense it doesn't mean anything or sound like anything else. So I don't know how I can improve this.

File: cql3/statements/create_keyspace_statement.cc (hunk 3 above)
Comment (nyh) [reply cql3/statements/create_keyspace_statement.cc:123]: Since people don't have markdown support on their cqlsh, I don't think it looks clearer, but perhaps us githubbers indeed got used to it, so I can do it.

File: cql3/statements/create_keyspace_statement.cc (hunk 5 above)
Comment (nyh) [reply cql3/statements/create_keyspace_statement.cc:120]: I think this message will be pointless if it doesn't explicitly list the specific features that have problems with tablets. I don't want to "cover my ass" and say something like "some things don't work with tablets, but I'm not going to commit on what doesn't work, so just watch your step". 
If we discover more problems, we should update this message.

File: cql3/statements/create_keyspace_statement.cc (hunk 2 above)
Comment (ptrsmrn) [reply cql3/statements/create_keyspace_statement.cc:119]: Yeah, me neither, I guess we have to learn to live with that.

**Review by ptrsmrn — APPROVED** [review]

**Comment by bhalevy** [comment]
@nyh is this PR ready for merge?

**Comment by bhalevy** [comment]
@scylladb/scylla-maint please consider merging

**Review by tgrabiec — APPROVED** [review]
